crates/compiler/src/output/declaration.rs

```diff
@@ -0,0 +1,239 @@
+
+    /// The default value of this declaration, if no value has been
+    /// specified in code or is available from a [`Dialogue`]'s
+    /// [`IVariableStorage`].
+    pub default_value: Value,
+
+    /// A string describing the purpose of this declaration.
+    pub description: String,
+
+    /// The name of the file in which this declaration was found.
+    ///
+    /// If this declaration was not found in a Yarn source file, this
+    /// will be [`DeclarationSource::External`].
+    pub source_file_name: DeclarationSource,
+
+    /// The name of the node in which this declaration was found.
+    ///
+    /// If this declaration was not found in a Yarn source file, this
+    /// will be [`None`].
+    pub source_node_name: Option<String>,
+
+    /// The line number at which this declaration was found in the
+    /// source file.
+    ///
+    /// If this declaration was not found in a Yarn source file, this
+    /// will be [`None`].
+    pub source_file_line: Option<usize>,
+
+    /// A value indicating whether this declaration was implicitly
+    /// inferred from usage.
+    ///
+    /// If `true`, this declaration was implicitly inferred from usage.
+    /// If `false`, this declaration appears in the source code.
+    pub is_implicit: bool,
+
+    /// The types of the variable, as represented by an object found
+    /// in a variant of [`DeclarationType`].
+    pub r#type: Type, // TODO
+
+    /// The range of text at which this declaration occurs.
+    ///
+    /// This range refers to the declaration of the symbol itself, and
+    /// not any syntax surrounding it. For example, the declaration
+    /// `<<declare $x = 1>>` would have a range referring to the `$x`
+    /// symbol.
+    pub range: RangeInclusive<Position>,
+}
+
+#[derive(Debug, Clone)]
+pub enum DeclarationType {}
+
+#[derive(Debug, Clone)]
+pub enum DeclarationSource {
+    External,
+    File(String),
+}
+
+/*
+/// <summary>
+/// Represents a position in a multi-line string.
+/// </summary>
+[System.Serializable]
+public class Position
+{
+    /// <summary>
+    /// Gets or sets the zero-indexed line of this position.
+    /// </summary>
+    public int Line { get; set; } = -1;
+
+    /// <summary>
+    /// Gets or sets the zero-indexed character number of this position.
+    /// </summary>
+    public int Character { get; set; } = -1;
+
+    /// <inheritdoc/>
+    public override bool Equals(object obj)
+    {
+        return obj is Position position &&
+               this.Line == position.Line &&
+               this.Character == position.Character;
+    }
+
+    /// <inheritdoc/>
+    public override int GetHashCode()
+    {
+        int hashCode = 1927683087;
+        hashCode = (hashCode * -1521134295) + this.Line.GetHashCode();
+        hashCode = (hashCode * -1521134295) + this.Character.GetHashCode();
+        return hashCode;
+    }
+}
+*/
+
+/// Represents a position in a multi-line string.
+#[derive(Debug, Clone)]
+pub struct Position {

```

Comment:
Curiosity: Why sometimes pub, and sometimes pub crate?

Reply:
Pub means this is intended to be public API, pub(crate) means this is only used internally. In theory, you could use pub everywhere and just not make the mod pub (access modifier are hierarchical, if the mod is private, all its items that are marked as pub are actually pub(crate)), but following this distinction now makes sure that we can use some more accurate lints, in particular dead code and missing docs for public API